protostuff-collectionschema/src/main/java/io/protostuff/MapSchema.java

```diff
@@ -47,6 +47,44 @@
         public Class<?> typeClass();
     }
 
+    private static final Comparable EMPTY_OBJECT = new Comparable()
+    {
+        @Override
+        public int compareTo(Object o)
+        {
+            return o == this ? 0 : -1;
+        }
+        @Override
+        public boolean equals(Object o)
+        {
+            return o == this;
+        }
+        @Override
+        public int hashCode()
+        {
+            return 0;
+        }
+    };
+
+    @SuppressWarnings("unchecked")
+    public static <T extends Map> T removeUnmodifiableWrapper(T map, MessageFactory messageFactory)
+    {
+        if(map == null)
+            return null;
+
+        try
+        {
+           map.put(EMPTY_OBJECT, EMPTY_OBJECT);
+           map.remove(EMPTY_OBJECT);
+        }

```

Comment:
Yes of cource, but the solution, in my opinion, is more robust than other solutions. 

I saw several ways to determine a unmodifiable collection:

1. Add/remove element and catch `UnsupportOperationException`
2. Compare types, for example:
```java 
List<String> list = new ArrayList<>();
List<String> unmodifiableList = Collections.unmodifiableList(list);
boolean isUnmodifiable = Collections.unmodifiableList(list).getClass()
                                                            .isInstance(unmodifiableList); 
```
3. It is the continuation of the second solution. Create list of types and compare with them.
```java  
for (String className: classNames)
       if (collection.getClass().getSimpleName().equals(className))
	      return true;
```
The second and third solutions is more elegant than first, but they are not flexible. We create list with unmodifiable collection types from `java.util.Collections.*` and we don't respect for any unmodifiable colections from other libraries.

I will glad, if you offer other solution for this problem.